bus-mapping/src/circuit_input_builder.rs

```diff
@@ -825,84 +1217,213 @@ impl<'a> CircuitInputBuilder {
+            if matches!(op.field, CallContextField::RwCounterEndOfReversion) {
+                let (tx_idx, call_idx) = self
+                    .block_ctx
+                    .call_map
+                    .get(&op.call_id)
+                    .expect("call_id not found in call_map");
+                op.value = self.block.txs[*tx_idx].calls[*call_idx]
+                    .rw_counter_end_of_reversion
+                    .into();
+            }
+        }
+    }
+
+    /// Handle a block by handling each transaction to generate all the
+    /// associated operations.
+    pub fn handle_block(
+        &mut self,
+        eth_block: &EthBlock,
+        geth_traces: &[eth_types::GethExecTrace],
+    ) -> Result<(), Error> {
+        for (tx_index, tx) in eth_block.transactions.iter().enumerate() {
+            let geth_trace = &geth_traces[tx_index];
+            self.handle_tx(tx, geth_trace, tx_index == eth_block.transactions.len() - 1)?;
+        }
+        self.set_value_ops_call_context_rwc_eor();
+        Ok(())
     }
 
     /// Handle a transaction with its corresponding execution trace to generate
     /// all the associated operations.  Each operation is registered in
     /// `self.block.container`, and each step stores the [`OperationRef`] to
     /// each of the generated operations.
-    pub fn handle_tx(
+    fn handle_tx(
         &mut self,
         eth_tx: &eth_types::Transaction,
         geth_trace: &GethExecTrace,
+        is_last_tx: bool,
     ) -> Result<(), Error> {
-        let mut tx = self.new_tx(eth_tx)?;
-        let mut tx_ctx = TransactionContext::new(eth_tx);
+        // Iterate over geth_trace to inspect and collect each call's
+        // is_success, which is at the top of stack at the step after a
+        // call.
+        let mut call_is_success = HashMap::new();
+        let mut call_indices = Vec::new();
+        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
+            if let Some(geth_next_step) = geth_trace.struct_logs.get(index + 1) {
+                // Dive into call
+                if geth_step.depth + 1 == geth_next_step.depth {
+                    call_indices.push(index);
+                // Emerge from call
+                } else if geth_step.depth - 1 == geth_next_step.depth {
+                    let is_success = !geth_next_step.stack.last()?.is_zero();
+                    call_is_success.insert(call_indices.pop().unwrap(), is_success);
+                }
+            }
+        }
+
+        let tx_is_success = !geth_trace.failed;
+        let mut tx = self.new_tx(eth_tx, tx_is_success)?;
+        let mut tx_ctx = TransactionContext::new(self.block.txs.len() + 1, is_last_tx, eth_tx);
+        self.block_ctx
+            .call_map
+            .insert(tx.calls[0].call_id, (self.block.txs.len(), 0));
+
+        // TODO: Move into gen_associated_steps with
+        // - execution_state: BeginTx
+        // - op: None
+        // Generate BeginTx step
```

Comment:
I think it will be moved to `gen_step` of its own kind in the future. At least to make it less verbose, I implement `Default` for `ExecStep`, do you think it's okay in this way?